### PR TITLE
[TC Automation] FADAVolMigrateValidation

### PR DESCRIPTION
**Pull request by spai-px**
What this PR does / why we need it:
This PR adds FADAVolMigrateValidation test case

Which issue(s) this PR fixes:
Closes # /PTX-22021

Special notes for your reviewer:
[jenkins](https://jenkins.pwx.dev.purestorage.com/job/Users/job/krishna/job/tp-pxe-byoc/1622/console)

### CODE REVIEW

File: tests/basic/pure_test.go [hunk 1]
```diff
@@ -2286,3 +2286,184 @@ func faLUNExists(faVolList []string, pvc string) bool {
...
+						}
+					}(ctx)
+				}
+				wg.Wait()
+			})
+
+			stepLog = "run the multipath -ll command on the node where the pods were scheduled before deleting"
+			Step(stepLog, func() {
+				// sleep for 60 seconds for all the entries to update
+				time.Sleep(60 * time.Second)
```
Comment (lsrinivas-pure) [inline tests/basic/pure_test.go:2412]: make it 30 secs

File: tests/basic/pure_test.go [hunk 2]
```diff
@@ -2286,3 +2286,184 @@ func faLUNExists(faVolList []string, pvc string) bool {
...
+			stepLog = "Check if the old multipath device entry is deleted from the node where the volume was attached"
+			Step(stepLog, func() {
+				//run the multipath -ll command on the node where the volume is attached
+				cmd := fmt.Sprintf("multipath -ll")
+				output, err := runCmd(cmd, selectedNode)
+				log.FailOnError(err, "Failed to run multipath -ll command on node %v", selectedNode.Name)
+				log.InfoD("Output of multipath -ll command: %v", output)
+				//check if the device path is present in multipath
+				for _, devicePath := range devicePaths {
+					if strings.Contains(output, devicePath) {
```
Comment (lsrinivas-pure) [inline tests/basic/pure_test.go:2450]: add check for failed faulty running

Comment (spai-px) [reply tests/basic/pure_test.go:2452]: Done

File: tests/basic/pure_test.go [hunk 3]
```diff
@@ -2166,3 +2166,196 @@ func faLUNExists(faVolList []string, pvc string) bool {
...
+				log.InfoD("Output of multipath on provisioned node -ll command: %v", output)
+				//check if the device path is present in multipath
+				if strings.Contains(output, "failed faulty running") {
+					log.FailOnError(fmt.Errorf("Multipath device error not detected"), "Multipath device error should be detected")
+				}
+				log.InfoD("Successfully validated that the old multipath device is deleted")
+			})
+			stepLog = "Destroy apps"
+			Step(stepLog, func() {
+				DestroyApps(contexts, nil)
```
Comment (kphalgun-px) [inline tests/basic/pure_test.go:2350]: log.InfoD?

Comment (spai-px) [reply tests/basic/pure_test.go:2351]: Done

File: tests/basic/pure_test.go [hunk 4]
```diff
@@ -2166,3 +2166,197 @@ func faLUNExists(faVolList []string, pvc string) bool {
 	}
 	return false
 }
+
+var _ = Describe("{FADAVolMigrateValidation}", func() {
+	var contexts []*scheduler.Context
+	JustBeforeEach(func() {
+		StartTorpedoTest("FADAVolMigrateValidation", "Attach FADA PVC on Node 1, confirm proper attachment. Stop PX on Node 1, ensure volume persistence in multipath -ll. Move deployment to Node 2, validate successful pod startup. Paths on original node indicate failure. Restart PX on Node 1, confirm old multipath device absence.", nil, 0)
```
Comment (lsrinivas-pure) [inline tests/basic/pure_test.go:2173]: make it a short text

File: tests/basic/pure_test.go [hunk 5]
```diff
@@ -2166,3 +2166,197 @@ func faLUNExists(faVolList []string, pvc string) bool {
...
+				}
+
+				StopVolDriverAndWait([]node.Node{selectedNode})
+			})
+			defer func() {
+				err = core.Instance().UnCordonNode(selectedNode.Name, defaultCommandTimeout, defaultCommandRetry)
+				log.FailOnError(err, "Failed to uncordon node %v", selectedNode.Name)
+				log.Infof("uncordoned node %v", selectedNode.Name)
+
+				err = Inst().S.RemoveLabelOnNode(secondNode, "apptype")
```
Comment (lsrinivas-pure) [inline tests/basic/pure_test.go:2244]: RemoveLabelOnNode is deferred twice I think . can u plz check

File: tests/basic/pure_test.go [hunk 6]
```diff
@@ -2166,3 +2166,197 @@ func faLUNExists(faVolList []string, pvc string) bool {
...
+				output, err := runCmd(cmd, selectedNode)
+				log.FailOnError(err, "Failed to run multipath -ll command on node %v", selectedNode.Name)
+				log.InfoD("Output of multipath -ll command: %v", output)
+				//check if the device path is present in multipath
+				for _, devicePath := range devicePaths {
+					if strings.Contains(output, devicePath) {
+						log.FailOnError(fmt.Errorf("Multipath device %v is still present", devicePath), "Multipath device %v should be deleted", devicePath)
+					}
+				}
+				output, err = runCmd(cmd, selectedNode)
```
Comment (lsrinivas-pure) [inline tests/basic/pure_test.go:2339]: why are you running twice. you already got the output. check the same

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

File: tests/basic/pure_test.go (hunk 4 above)
Comment (spai-px) [reply tests/basic/pure_test.go]: Done. added the description as a comment

File: tests/basic/pure_test.go (hunk 6 above)
Comment (spai-px) [reply tests/basic/pure_test.go]: Done

File: tests/basic/pure_test.go (hunk 5 above)
Comment (spai-px) [reply tests/basic/pure_test.go:2252]: This removelabel is for the second node where the apps are moved to. so this is required.

File: tests/basic/pure_test.go [hunk 7]
```diff
@@ -2166,3 +2166,202 @@ func faLUNExists(faVolList []string, pvc string) bool {
...
+	/* 
+          1. Attach FADA PVC on Node 1, confirm proper attachment. 
+	  2. Stop PX on Node 1, ensure volume persistence in multipath -ll. 
+          3. Move deployment to Node 2, validate successful pod startup. 
+	  4. Paths on original node indicate failure. Restart PX on Node 1, confirm old multipath device absence.
+
+        */
+	var contexts []*scheduler.Context
+	JustBeforeEach(func() {
+		StartTorpedoTest("FADAVolMigrateValidation", "Migrate pods from node 1 to node and check multipath consistency", nil, 0)
```
Comment (kphalgun-px) [inline tests/basic/pure_test.go:2181]: How about? or something similar

~node 1 to node~ one node to another node?

**Review by lsrinivas-pure — APPROVED** [review]
LGTM

**Pull request merged**